### PR TITLE
Start moving client logic to nested service layers

### PR DESCRIPTION
**Pull request by sfackler**
## Before this PR
The bulk of the client logic was open-coded, making it hard to test various bits of behavior in isolation without just running an HTTP server. It also makes it harder to understand the specific behavior of the client in some situations.

## After this PR
Following dialogue's lead, this PR starts moving logic out to multiple `Service` types, somewhat analogous to dialogus's `Channel` interface. Each handles a specific bit of logic, and are nested together in the main client.

Only a few bits of the "innermost" bits of logic have been ported so far, but it should give an indication of what this'll look like moving forward. The total line count increases, but much of that is due to us actually testing the individual components now!

cc #18

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -5,3 +5,6 @@ members = [
     "conjure-verification",
     "conjure-verification-api",
 ]
+
+[patch.crates-io]
```
Comment (sfackler) [inline Cargo.toml:9]: We're depending on some hyper changes that haven't released yet, but I expect that will be resolved by the time we finish porting everything.

File: conjure-runtime/src/service/gzip.rs [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+        self.project()
+            .stream
+            .get_pin_mut()
+            .project()
+            .body
+            .poll_trailers(cx)
+            .map_err(|e| io::Error::new(io::ErrorKind::Other, e))
+    }
+
+    // we can't override is_end_stream since we may get an error out of the gzip decoder at eof
```
Comment (jonsyu1) [inline conjure-runtime/src/service/gzip.rs:244]: Is this the best place for this comment? How about above [line 216](https://github.com/palantir/conjure-rust-runtime/pull/32/files#diff-17d875a42152ea09fa8c11d259b4dc0eR216)?

File: conjure-runtime/src/service/proxy/request.rs [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+use tower::layer::Layer;
+use tower::Service;
+use url::Url;
+
+/// A layer which adjusts an HTTP request as necessary to respect proxy settings.
+///
+/// This is paired with the `ProxyConnectorLayer` which handles the socket-level proxy configuration.
+///
+/// For http requests over an HTTP proxy, we inject the `Proxy-Authorization` header if necessary. For https requests
+/// authorization is handled in the `CONNECT` request made by the `ProxyCOnnectorLayer`.
```
Comment (jonsyu1) [inline conjure-runtime/src/service/proxy/request.rs:29]:
```suggestion
/// authorization is handled in the `CONNECT` request made by the `ProxyConnectorLayer`.
```

**Review by jonsyu1 — APPROVED** [review]
LGTM

File: conjure-runtime/src/service/gzip.rs (hunk 2 above)
Comment (sfackler) [reply conjure-runtime/src/service/gzip.rs:244]: I stuck it here since that's where the override "would go" but the top of the impl works too.

**Pull request merged**